### DISCUSSION TITLE
LDM Notes for December 11th, 2023

### DISCUSSION THREAD

**Discussion by 333fred** [dotnet/csharplang#7777] [LDM Notes]
https://github.com/dotnet/csharplang/blob/main/meetings/2023/LDM-2023-12-11.md

## Agenda

- Extensions

**Comment by TahirAhmadov** [comment]
What happens if two in scope implicit extensions for the same type implement the same interface? Currently for extension methods it measures namespace proximity and errors if both are equally close - can the same approach be used here?

For implicit to implicit inheritance, would it help if we were to allow defining `protected` members on the extension itself, and then child extension can access base extension members?

PS. Speaking of - can extensions define `private` members? This would be purely for code reuse/modularization.
PPS. I just re-read the [proposal](https://github.com/dotnet/csharplang/blob/main/proposals/extensions.md) - yes all access modifiers are allowed. So the second question above is pertinent.

**Reply by CyrusNajmabadi** [reply]
> Currently for extension methods it measures namespace proximity 

I don't think there is any measurement of proximity.  Instance methods are looked up first.  Then extensions.  All extensions from all the imported usings are considered together.

**Reply by TahirAhmadov** [reply]
Please look at [this](https://sharplab.io/#v2:EYLgtghglgdgNAFxAJwK7wCYgNQB8ACATAAwCwAUEQIwUUwRgCmAzgA4QDGjABAMIUBvCtxHciYqgHZhooeVEKJANjEAWbgFloMABQBKGYu5yjR/FQCcOgESAeDcAg+9b0BuQ6ZFUAdADEA9r/1XeVMAXzcw4JF8AGZlMUJuAEE3E1MYuPx1PwCEAAsoZm5YBG4AFXzmPWMIhRrRdPMVcUTCFLcFBqomrP8dPIKimBLygqqBOpEIiM6mhIAhQTcZtW5svorB4YqxqYogA===).
If you comment out `A2`, it works - choosing `A` over `B`, because `A` is "closer".

**Reply by CyrusNajmabadi** [reply]
this is not an extension brought in by a using.  That's just lookup within your own namespaces.  That's the much rarer case, but certainly does exist :)

We intend to have our extension lookup follow the same rules.

**Comment by huoyaoyuan** [comment]
Have the team explored the option to do a massive change on CLR type system to support higher level concepts?
